### ISSUE TITLE
Type guard and function call

### ISSUE THREAD

**Issue by maximelkin** [microsoft/TypeScript#26729]
**TypeScript Version:**  3.1.0-dev.201xxxxx

**Search Terms:** typeguard type guard function union nested

**Code**

```ts
enum Test {
    A,
    B,
    C
};

let a: { readonly prop: Test } = { prop: Test.A };

if (a.prop === Test.A) {
    /* Error:
    Argument of type '{ readonly prop: Test; }' is not assignable to parameter of type '{ readonly prop: Test.A | Test.B; }'.
  Types of property 'prop' are incompatible.
    Type 'Test' is not assignable to type 'Test.A | Test.B'.
     */
    f(a);
}

function f(arg: {readonly prop: Test.A | Test.B}) {
}
```

**Expected behavior:** no error

**Actual behavior:** error on call f

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=enum%20Test%20%7B%0D%0A%20%20%20%20A%2C%0D%0A%20%20%20%20B%2C%0D%0A%20%20%20%20C%0D%0A%7D%3B%0D%0A%0D%0Alet%20a%3A%20%7B%20readonly%20prop%3A%20Test%20%7D%20%3D%20%7B%20prop%3A%20Test.A%20%7D%3B%0D%0A%0D%0Aif%20(a.prop%20%3D%3D%3D%20Test.A)%20%7B%0D%0A%20%20%20%20%2F*%0D%0A%20%20%20%20Argument%20of%20type%20'%7B%20readonly%20prop%3A%20Test%3B%20%7D'%20is%20not%20assignable%20to%20parameter%20of%20type%20'%7B%20readonly%20prop%3A%20Test.A%20%7C%20Test.B%3B%20%7D'.%0D%0A%20%20Types%20of%20property%20'prop'%20are%20incompatible.%0D%0A%20%20%20%20Type%20'Test'%20is%20not%20assignable%20to%20type%20'Test.A%20%7C%20Test.B'.%0D%0A%20%20%20%20%20*%2F%0D%0A%20%20%20%20f(a)%3B%0D%0A%7D%0D%0A%0D%0Afunction%20f(arg%3A%20%7Breadonly%20prop%3A%20Test.A%20%7C%20Test.B%7D)%20%7B%0D%0A%7D

**Comment by maximelkin** [comment]
Copy of https://github.com/Microsoft/TypeScript/issues/14350

**Comment by RyanCavanaugh** [comment]
Object types containing a single property of a union type aren't special-cased to be equivalent to an expanded form of themselves.

If you add additional properties here you can see why narrowing in the `else` branch isn't sound in the multi-property case.
